File: day:56/zigzagLevelOrder.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def zigzagLevelOrder(self, root: TreeNode) -> List[List[int]]:
        if(root is None): return [];
    
        stack = [root];
        result = [];
        left = True;
    
        while(len(stack) > 0):
            newStack = [];
            temp = [];
            while(len(stack) > 0):
                node = stack.pop(0);
                temp.append(node.val);
                if(node.left):
                    newStack.append(node.left);
                if(node.right):
                    newStack.append(node.right);
            if(left):
                result.append(temp); 
            else:
                result.append(list(reversed(temp)))
        
            left = not left; 
            stack = newStack;
        return result;
    

File: day:56/zigzagLevelOrder.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var zigzagLevelOrder = function(root) {
    if(root === null) return [];
    
    let stack = [root];
    let result = [];
    let left = true;
    
    while(stack.length > 0) {
        let newStack = [];
        let temp = [];
        let node;
        while(stack.length > 0) {
            node = stack.shift();
            temp.push(node.val);
            if(node.left) {
                newStack.push(node.left);
            }
            if(node.right) {
                newStack.push(node.right);
            }
        }
        if(left) {
           result.push(temp); 
        }
        else {
            result.push(temp.reverse())
        }
        
        left = !left; // change the direction!
        stack = newStack;
    }
    return result;
    
};
